docs/configuring_keyword_alignment.rst

```diff
@@ -241,11 +244,54 @@ Rule Specific Keyword Alignment Configuration
           when others =>
               null;
       end case
+      data_valid_after <= '1';
 
-    **Fix (case_control_statements_end_group = False)**
+    **Fix (case_control_statements_ends_group = False)**
 
     .. code-block:: vhdl
 
+      data_valid_before <= '1';
+      case A is
+          when A =>
+              X         <= F;
+              XY        <= G;
+              XYZ       <= H;
+          when B =>
+              a         <= I;
+              ab        <= h;
+              c         <= a;
+          when others =>
+              null;
+      end case
+      data_valid_after  <= '1';
+
+#. :code:`case_keyword_statements_ends_group` - if set to :code:`True` and :code:`case_control_statements_ends_group` is set to :code:`False`, any line with case when statement ends the group of lines that should be aligned and starts new group. If set to :code:`False`, this configuration has no effect and :code:`case_control_statements_ends_group` rules are followed. 
```

Comment:
So if I understand this correctly, if `case_control_statements_ends_group` is set to `True` it would stop aligning when it hit the `case` keyword.  However, you could not control alignment across the `when` keywords.

When `case_control_statements_ends_group` is set to `False` then alignment would occur through the case statement.  Essentially aligning everything under all the `when` statements with everything outside the `case` statement.

With the new `case_keyword_statements_ends_group` set to `True` then the statements under the `when` keywords will not be aligned.

I feel like I missed something.

Reply:
The idea of the new `case_keyword` attribute is:
* if left at a default of False, the existing behavior of the `case_control` attribute is maintained for backwards compatibility
* this existing behavior is:
   * True means the `case/end case` AND* `when` keywords ends a group (*important)
   * False means that NEITHER `case/end case` NOR `when` ends a group
 * so the only new behavior happens when `case_keyword` is True and `case_control` is False. This means that `case/end case` ends a group (from the `case_keyword`) but `when` does NOT end a group (from the `case_control`)
 * the new attribute allows `when` statements to be configured differently to `case/end case` statements
 * I added some assignments above and below the case statements to try and show what happens outside the case statements

Maybe take another look at the docs for this, and feel free to improve the explanation - I found it hard!

Reply:
k, I think I understand what you were going for.  Would you be up for either of these changes:

Combine options
============

  1)  remove `case_keyword_statements_ends_group`
  2)  add third option to `case_control_statements_ends_group`:  `True`, `False`, `break_at_when`

Change new option
=============

  1)  rename `case_keyword_statements_ends_group` to `case_when_keyword_ends_group`
  2)  This would perform the same effect, but would only take affect if `case_control_statements_ends_group` is `True`

I think that might clear up the documentation, because `case_control_statements_ends_group` would not be counter intuitive.

Reply:
Before I look at your suggestions, I've just realised that the docs were wrong and I've made a correction. Please could you take a look at the new docs and then let me know how you want to proceed - happy to continue with your options once I'm sure you've accounted for the doc error I've corrected....

Reply:
I've looked at your options and I think there is still a slight misunderstanding...

In the combine options change, the third option would have to be something like ends_at_case_and_case_end.

I cannot see the advantage of the `change new option` idea as it stands

Reply:
I pulled down your pull request and I am working on an example so I can understand how the options work.

So given this example code:
```vhdl
  1
  2 architecture rtl of fifo is
  3
  4 begin
  5
  6   label : process () is
  7   begin
  8
  9     a <= 1;
 10
 11     b_blah <= 2;
 12
 13     case blah is
 14
 15       when 0000 =>
 16
 17         a <= 1;
 18         b_blah <= 2;
 19
 20       when 0010 =>
 21
 22         a <= 2;
 23         b_blah2 <= 3;
 24
 25     end case;
 26
 27     a <= 1;
 28
 29     b_blah <= 2;
 30
 31   end process label;
 32
 33 end architecture rtl;
```
If I use this configuration:
```yaml
  1 rule:
  2   sequential_005:
  3     blank_line_ends_group: False
  4     comment_line_ends_group: True
  5     if_control_statements_ends_group: True
  6     case_control_statements_ends_group: False
  7     case_keyword_statements_ends_group: True
  8     loop_control_statements_ends_group: False
```
Results in the following:
```vhdl
  1
  2 architecture rtl of fifo is
  3
  4 begin
  5
  6   label : process () is
  7   begin
  8
  9     a      <= 1;
 10
 11     b_blah <= 2;
 12
 13     case blah is
 14
 15       when 0000 =>
 16
 17         a       <= 1;
 18         b_blah  <= 2;
 19
 20       when 0010 =>
 21
 22         a       <= 2;
 23         b_blah2 <= 3;
 24
 25     end case;
 26
 27     a      <= 1;
 28
 29     b_blah <= 2;
 30
 31   end process label;
 32
 33 end architecture rtl;
```
I would have expected the `=>` on lines 17 and 18 to be one column to the left.
With this configuration:
```yaml
  1 rule:
  2   sequential_005:
  3     blank_line_ends_group: False
  4     comment_line_ends_group: True
  5     if_control_statements_ends_group: True
  6     case_control_statements_ends_group: True
  7     case_keyword_statements_ends_group: False
  8     loop_control_statements_ends_group: False
```
results in:
```vhdl
 13     case blah is
 14
 15       when 0000 =>
 16
 17         a      <= 1;
 18         b_blah <= 2;
 19
 20       when 0010 =>
 21
 22         a       <= 2;
 23         b_blah2 <= 3;
 24
 25     end case;
```
Is that backwards from what you were intending?~

Reply:
To see the third behavior, try this:

  1 rule:
  2   sequential_005:
  3     blank_line_ends_group: False
  4     comment_line_ends_group: True
  5     if_control_statements_ends_group: True
  6     case_control_statements_ends_group: False
  7     case_keyword_statements_ends_group: False
  8     loop_control_statements_ends_group: False

Reply:
Let's number the three examples in order: 1,2 (and 3 if you try my suggestion)

In example 1, the groups end on lines 12 and 25 only

In example 2, the groups end on lines 12, 15, 20 and 25

In example 3, when you've done it, there will be no groups end!

Examples 2 and 3 are existing behavior, example 1 is the new one I want to be able to achieve

Reply:
To answer you question, its working as I intended :-)

Reply:
It's slightly hard to achieve what I want and keep backwards compatibility, which is essential I believe. I think what I've done works, it's just coming up with a good name etc

Reply:
Example 1
-----------
```yaml
rule:
  sequential_005:
    blank_line_ends_group: False
    case_control_statements_ends_group: False
    case_keyword_statements_ends_group: False
```
```vhdl
 15       when 0000 =>
 16
 17         a       <= 1;
 18         b_blah  <= 2;
 19
 20       when 0010 =>
 21
 22         a       <= 2;
 23         b_blah2 <= 3;
```
Where it is aligned with lines 9, 11, 27 and 29.
Where lines 9, 11, 17, 18, 22, 23, 27 and 29 are aligned.

Example 2
------------
```yaml
rule:
  sequential_005:
    blank_line_ends_group: False
    case_control_statements_ends_group: True
    case_keyword_statements_ends_group: False
```
```vhdl
 15       when 0000 =>
 16
 17         a      <= 1;
 18         b_blah <= 2;
 19
 20       when 0010 =>
 21
 22         a       <= 2;
 23         b_blah2 <= 3;
```
Where lines 9, 11, 27 and 39 are not aligned with 17, 19, 22 and 23.
Where lines 17 and 18 are aligned.
Where lines 22 and 23 are aligned.
Where lines 17 and 18 are not aligned with 22 and 23.

Example 3
------------
```yaml
rule:
  sequential_005:
    blank_line_ends_group: False
    case_control_statements_ends_group: False
    case_keyword_statements_ends_group: True
```
```vhdl
 15       when 0000 =>
 16
 17         a       <= 1;
 18         b_blah  <= 2;
 19
 20       when 0010 =>
 21
 22         a       <= 2;
 23         b_blah2 <= 3;
```
Where lines 17, 18, 22 and 23 are aligned.

okay....I think I get it now.

`case_control_statements_ends_group` will create groups on `case`, `end case`, and `when` keywords.
`case_keyword_statements_ends_group` will create groups on `case` and `end case` keywords only.

Is my understanding aligned with your intent?

Reply:
I believe so
* now you have seen the intent, please suggest better ways of configuring this if you can. I quite like making case_control a 3 way flag
* also the PR code coverage needs looking at...again feel free to suggest improvements as my Python is not expert

Reply:
> now you have seen the intent, please suggest better ways of configuring this if you can. I quite like making case_control a 3 way flag

I will add some comments to the `align_tokens_in_region_between_tokens_unless_between_tokens.py` file.